Clarifying how sharding works (#6853)
hammerandchisel/discord-api-docs: docs/topics/Gateway.md

## Sharding

> As apps grow and are added to an increasing number of guilds, some developers may find it necessary to divide portions of their app's operations across multiple processes. As such, the Gateway implements a method of user-controlled guild sharding which allows apps to split events across a number of Gateway sessions. Guild sharding is entirely controlled by an app, and requires no state-sharing between separate sessions to operate. While all apps *can* enable sharding, it's not necessary for apps in a smaller number of guilds.

Reviewer: We don't really have something called a "gateway session" but I understand the intention here.

To elaborate on the terminology, a gateway connection refers to a connection to our websocket gateway at gateway.discord.gg, and a gateway connection then spawns a session, or re-establishes a connection to an existing session. The session outlives the gateway connection, since you can re-connect to the gateway when you're disconnected, and RESUME to re-establish the gateway socket's connection to a given session.

Author: So it should just be "session" instead of "Gateway session"?

Reviewer: Yeah that would be fine.

Reviewer 2: Is session more accurate than connection here? IMO connections are more intuitive than sessions, so rewriting this section in terms of sessions makes it harder to understand.

Author: I believe that session is more accurate yes.
As Jake wrote, a session can be RESUMED in a new connection to the gateway. And, as per the [documentation](https://discord.com/developers/docs/topics/gateway#resuming), a connection will be sent all missed events from a session once it resumes it.
Thus, events being sent to a session which "forwards" them over the active connection or stores them if a connection is not currently active is a more accurate way of thinking of it (unless I completely misunderstand how it works).

> warn
> Each shard can only support a maximum of 2500 guilds, and apps that are in 2500+ guilds *must* enable sharding.

> Sessions that would like to only receive events from a subset of guilds should send the `shard` array in the [Identify](#DOCS_TOPICS_GATEWAY_EVENTS/identify) payload. The first item in this array is `shard_id`, the zero-based integer value of the current shard, while the second is `total_shards` and represents the total number of shards.

Reviewer 3: ```suggestion
Sessions that would like to only receive events from a subset of guilds should send the `shard` array in the [Identify](#DOCS_TOPICS_GATEWAY_EVENTS/identify) payload. The first item in this array is `shard_id`, the zero-based integer value of the current shard, while the second is `num_shards` and represents the total number of shards.
```
For consistency with the remaining of the page.

> info
> The [Get Gateway Bot](#DOCS_TOPICS_GATEWAY/get-gateway-bot) endpoint provides a recommended number of shards for your app in the `shards` field

> A certain gateway session is only subscribed to events from guilds with a `guild_id` that fulfil the following formula, using the `shard_id` and `num_shards` that the session provided in the Identify event:

Reviewer 3: ```suggestion
A certain gateway session is only subscribed to events from guilds with a `guild_id` that fulfil the following formula, using the `shard_id` and `num_shards` that the session provided in the [Identify](#DOCS_TOPICS_GATEWAY_EVENTS/identify) event:
```

> A certain gateway session is only subscribed to events from guilds with a `guild_id` that fulfil the following formula, using the `shard_id` and `num_shards` that the session provided in the [Identify](#DOCS_TOPICS_GATEWAY_EVENTS/identify) event:

Reviewer 4: ```suggestion
A certain gateway session is only subscribed to events from guilds with a `guild_id` that satisfies the following formula, using the `shard_id` and `num_shards` that the session provided in the [Identify](#DOCS_TOPICS_GATEWAY_EVENTS/identify) event:
```
it would be "fulfils" but satisfies works better here

Author: I agree that "satisfy" is a better word choice, singular form is also correct thank you.

###### Sharding Formula

```python

> shard_id == (guild_id >> 22) % num_shards

Reviewer 5: ```suggestion
(guild_id >> 22) % num_shards == shard_id
```

that change is incorrect. `==` is an equality operator, `=` sets the `shard_id`, meaning it was correct.
This example is meant to calculate the shard id for given `guild_id`, not to check the equality.
This shouldn't be a change imo.
Libraries also use that forumula do calculate where guilds are

Reviewer 6: zoddo said that it is fine

![discussion](https://github.com/discord/discord-api-docs/assets/101461652/2b644422-977d-4fa7-9c89-82d94c083bf7)
```
- Hampus — Today at 18:34
also noticed this @Zoddo — seems like an incorrect find & replace, guess you could suggest an edit for that as well while you're at it
[<Image>](https://cdn.discordapp.com/attachments/381887113391505410/1238514447635582977/image.png?ex=663f8ff6&is=663e3e76&hm=f3f3010b08a21dfc6a2f774b18532fb1243bcc2e82ff640344af400ff87a0ba7&)

- Zoddo — Today at 18:35
no, I guess they did it on purpose. From the thread on ddevs:
[<Image>](https://cdn.discordapp.com/attachments/381887113391505410/1238514798854148146/image.png?ex=663f904a&is=663e3eca&hm=97d2d08a01ff346c5e0b436d449e94165958852399057e3c5fd3b5b864251af5&)
depending how you explain things, this makes more sense

- Hampus — Today at 18:35
huh
I guess — that seems more confusing to me though, `shard_id` is a variable, isn't it?

- Zoddo — Today at 18:40
yeah, but they "reversed" the explanation. Previously, the docs were explaining how to calculate which shard will receive an event, which seemed to cause difficulties to understand for non-initiated people. That's the initial question in the thread:
It says that you can start multiple shards with the same `shard_id` or start shards with different `num_shards`.
Well which num_shards is Discord using in the formula?
What happens when you have multiple shards with the same `shard_id`, how is that handled?

In their PR, the reformulated it the other way to explain which events a specific shard will receive.
I'm fine with both ways :blobshrug:

- Hampus — Today at 18:42
yeah, I don't really mind either, just think that it could be seen as a typo since you don't get a `shard_id` variable in scope to be referenced throughout the text by checking if it's equal to the right-hand side — but I see where they're coming from

- Zoddo — Today at 18:42
It's referenced literally just before the code block:
A certain gateway session is only subscribed to events from guilds with a `guild_id` that fulfil the following formula, using the `shard_id` and `num_shards` that the session provided in the Identify event:
the formula should be interpreted is in the context of a shard, so the `shard_id` is know

- Hampus — Today at 18:43
I see
did not notice that change (a review comment block was taking up space above it while I was viewing the changes, my bad
```

Reviewer 5: https://github.com/Pycord-Development/pycord/blob/08370b57368e2e5ce0e4e70faf8514742f45f5e5/discord/guild.py#L1011-L1016

https://github.com/DSharpPlus/DSharpPlus/blob/cc238ce9542ab1ef2137dd4816a6062cb4ea933c/DSharpPlus/Utilities.cs#L196-L197

https://github.com/discordjs/discord.js/blob/e673b3c129f288f9f271e0b991d16dc2901cdc8a/packages/util/src/functions/calculateShardId.ts#L7-L9

Reviewer 5: The most common usage is imo calculating based on the guild id, on which shard the guild is located on, hence I suggest not changing it

Reviewer 3: > This example is meant to calculate the shard id for given guild_id, not to check the equality.

No, the whole section preceding this formula explains which events will be received by a specific shard *(and not which shards will receive a specific event)*. So you should look at the formula in the context of a shard (the shard_id is already known in this case). The previous version of this section (which was centered around the guild instead of around the shard) was causing confusion around edge cases to devs that aren't proficient with how Discord is working; you should [take a look at the thread](https://discord.com/channels/613425648685547541/1208012213491736576/1208012213491736576) that ended up in the creation of this PR.

Anyway, your point about how libs are using this formula doesn't make sense: the goal of the documentation, especially for something that may be complex to understand, is to explain how this works, not to have code and formulas people can just copy&paste without thinking.

However, I agree this may interpreted as a typo, so to make it clear this is an equation, I suggest to do this:
```suggestion
(guild_id >> 22) % num_shards == shard_id
```

Reviewer 5: that sounds like a good way

Author: That seems reasonable to show that it is not a typo. Thank you.

> (guild_id >> 22) % num_shards == shard_id

Reviewer: I'm not entirely sure why this change is necessary - both say the same thing, and I think the prior version says it a bit more clearly.

Author: Do you mean whether `shard_id` is first or last, or whether it's an assignment `=` or comparison `==`?

In the first case, it is to less ambiguously show that it is, in fact, intended to be a comparison and not an assignment.

The whole point of this PR is to rewrite the documentation in terms of which of the guilds that a bot is in to which a gateway "session" with a certain `shard_array` will be subscribed to events, as opposed to deciding which "`shard_id` a certain event will be sent to" as this is very confusing in the context of multiple "sessions" with the same `shard_id` and different `num_shards`. See the original discussion that led to this PR and [Zoddos reply](https://github.com/discord/discord-api-docs/pull/6853#discussion_r1597226734) in Lulalaby's review above.

Reviewer: I think that `a = b + c` is much more clear than `b + c == a`

Author: With this PR, the formula is written in the context of - when a shard is connecting to the gateway - iterating through all guilds that a bot is part of and deciding whether the shard should receive events from that guild based on the guilds `guild_id` and the `shard` array with `shard_id` and `num_shards` that the shard provided.
Sure, if you see it as a mathematical equation, a single `=` would suffice, but this is documentation that programmers will be reading to understand how sharding behaves, and when programmers see an expression like that with things that are very clearly variable names they would assume that `=` is an assignment, which this formula is not.
I agree that writing the single variable on the left-hand side looks cleaner, but as discussed in the thread I mentioned, that would create _some_ ambiguity of whether it was actually meant as a comparison or not.

Reviewer 7: How is this not an assignment? The original wording is much clearer.

Reviewer: Although I appreciate the discussion, I would still prefer the original form to remain.

Author: Sure, we can keep it as an assignment. It still has to be in the context of a single shard though, as each session has its own (potentially different) value of `num_shards`. Then we can just write with words afterwards that "The session will receive events from every guild that, using the formula above, evaluates to the sessions `shard_id` using the corresponding `num_shards`".
I don't see how adding this level of indirection makes it any clearer though?

Reviewer: I think that simply keeping it as is for the time being is fine. I feel the formula is clear enough and one can grok the relationship. 

I've always thought the larger meaning here is: "given a guild, and the number of shards I'm running, which shard does a given guild land on". This answers a practical question, for example if you want to locate the shard that is handling a given guild, so you can look up data from that shard's process. Internally, we do such lookups (albeit with hash rings, but they are logically the same.). This is how the original formula was written. 

The, "which guilds precisely will land on this shard" is a less useful question to answer, as it's not something the developer has control over, and is a fact that one can derive from the original formula as well.

```

Every session with `shard_id = 0` will be subscribed to DM's and other non-guild related events.

> As an example, if you wanted to split events equally between three shards, you'd use the following values for `shard` for each session: `[0, 3]`, `[1, 3]`, and `[2, 3]`. DM's would only be sent to the `[0, 3]` shard.

Reviewer 3: ```suggestion
Every session with `shard_id = 0` will be subscribed to DMs and other non-guild related events.

As an example, if you wanted to split events equally between three shards, you'd use the following values for `shard` for each session: `[0, 3]`, `[1, 3]`, and `[2, 3]`. DMs would only be sent to the `[0, 3]` shard.
```

> As an example, if you wanted to split events equally between three shards, you'd use the following values for `shard` for each session: `[0, 3]`, `[1, 3]`, and `[2, 3]`. DMs would only be sent to the `[0, 3]` shard.

Reviewer: The word "evenly" is not necessarily true, since the volume of events dispatched to a given session is entirely dependent on the guilds on that shard. 

E.g. in a pathological case, if your bot is in 2 guilds, and you have one of the guilds has 1 member, and the other 500,000 members, you definitely won't see an "even" distribution of events.

Author: Good point. "As an example, if you wanted to split guilds equally between three shards" would probably be better wording, along with a remark that this of course won't necessarily split the events evenly if some guilds produce more events than others.
I'll include this in a batch once all your comments have been resolved.

Reviewer: There is no "split equally" either - for example, if you are running 3 shards, and you leave all guilds on shard 0, then your guilds are not and will no longer be split equally, and re-connecting would of course not repair this bias.

Reviewer: The entire basis of this system is that it takes a probabilistic approach to distributing guilds between shards, based on the millisecond that the guild was created in our system. With enough guilds and shards, it should balance out, but there definitely is no guarantee of an even or equal split one way or another.

Author: Yes, I understand that very well. There is an implicit assumption in that sentence that the bot is part of many guilds (as it would most likely be when sharding becomes necessary). I could definitely make that assumption explicit.

> Note that `num_shards` does not relate to (or limit) the total number of potential sessions, and can be different between multiple sessions existing at the same time. It is only used to decide whether an event will be sent to the associated session using the [Sharding Formula](#DOCS_TOPICS_GATEWAY/sharding-sharding-formula) above. In the simple case like the example above, where every session has the same `num_shards` and the sessions respective `shard_id`'s cover every value from `0` to `num_shards - 1`, the events will be split evenly between the sessions. This is probably how most bots will operate.

Reviewer: I would remove the word "evenly" here as well.

Author: Same as above, I agree.
